six/modules/c++/cphd/source/CPHDXMLControl.cpp

```diff
@@ -35,483 +43,998 @@ typedef xml::lite::Element* XMLElem;
+        {
+            throw six::DESValidationException(Ctxt(
+                "INVALID XML: URI is empty so document version cannot be "
+                "determined to use for validation"));
+        }
+
+        validator.validate(doc->getRootElement(),
+                           doc->getRootElement()->getUri(),
+                           errors);
+
+        // log any error found and throw
+        if (!errors.empty())
+        {
+            for (size_t i = 0; i < errors.size(); ++i)
+            {
+                log->critical(errors[i].toString());
+            }
+
+            //! this is a unique error thrown only in this location --
+            //  if the user wants a file written regardless of the consequences
+            //  they can catch this error, clear the vector and SIX_SCHEMA_PATH
+            //  and attempt to rewrite the file. Continuing in this manner is
+            //  highly discouraged
+            for (size_t i = 0; i < errors.size(); ++i)
+            {
+                std::cout << errors[i].toString();
+            }
+            throw six::DESValidationException(Ctxt(
+                "INVALID XML: Check both the XML being " \
+                "produced and the schemas available"));
+        }
+    }
 }
 
 std::string CPHDXMLControl::toXMLString(const Metadata& metadata)
 {
     std::auto_ptr<xml::lite::Document> doc(toXML(metadata));
     io::StringStream ss;
-    doc->getRootElement()->print(ss);
+    doc->getRootElement()->prettyPrint(ss);
 
     return (std::string("<?xml version=\"1.0\"?>") + ss.stream().str());
 }
 
-size_t CPHDXMLControl::getXMLsize(const Metadata& metadata)
-{
-    return toXMLString(metadata).size();
-}
-
+/*
+ * TO XML
+ */
 std::auto_ptr<xml::lite::Document> CPHDXMLControl::toXML(const Metadata& metadata)
 {
     std::auto_ptr<xml::lite::Document> doc(new xml::lite::Document());
 
     XMLElem root = newElement("CPHD");
     doc->setRootElement(root);
 
-    // Fill in the rest...
-    toXML(metadata.collectionInformation, root);
-    toXML(metadata.data, root);
+    toXML(metadata.collectionID, root);
     toXML(metadata.global, root);
+    toXML(metadata.sceneCoordinates, root);
+    toXML(metadata.data, root);
     toXML(metadata.channel, root);
-    toXML(metadata.srp, root);
+    toXML(metadata.pvp, root);
+    if (metadata.supportArray.get())
+    {
+        toXML(*(metadata.supportArray), root);
+    }
+    toXML(metadata.dwell, root);
+    toXML(metadata.referenceGeometry, root);
     if (metadata.antenna.get())
     {
-        toXML(*metadata.antenna, root);
+        toXML(*(metadata.antenna), root);
+    }
+    if (metadata.txRcv.get())
+    {
+        toXML(*(metadata.txRcv), root);
+    }
+    if (metadata.errorParameters.get())
+    {
+        toXML(*(metadata.errorParameters), root);
+    }
+    if (metadata.productInfo.get())
+    {
+        toXML(*(metadata.productInfo), root);
+    }
+    for (size_t i = 0; i < metadata.geoInfo.size(); ++i)
```

Comment:
Avoid single-character variable names (so `i` can be `ii`). It's mainly just convention, but there are also arguments about making the name easier to search for.